### PR TITLE
Some initial inference engine refactors for enabling training

### PR DESCRIPTION
**Pull request by blindcrone**
Only on MLX for now, breaks Tinygrad (doesn't fulfill interface yet)

### CODE REVIEW

File: exo/inference/inference_engine.py [hunk 1]
```diff
@@ -9,13 +9,23 @@
 
 class InferenceEngine(ABC):
   @abstractmethod
-  async def infer_prompt(self, request_id: str, shard: Shard, prompt: str, image_str: Optional[str] = None, inference_state: Optional[str] = None) -> Tuple[np.ndarray, str, bool]:
+  async def encode(self, shard: Shard, prompt: str) -> np.ndarray:
+    pass
+  
+  async def sample(self, x: np.ndarray) -> np.ndarray:
```
Comment (dtnewman) [inline exo/inference/inference_engine.py:15]: Should this be an abstract method?

Comment (blindcrone) [reply exo/inference/inference_engine.py:15]: O yup, sure should

**Comment by blindcrone** [comment]
Okay this no longer breaks tinygrad afaict

File: exo/inference/mlx/sharded_model.py [hunk 2]
```diff
@@ -87,3 +29,15 @@ def init_cache(self, request_id: str):
       self.caches.popitem(last=False)
 
     self.caches[request_id] = cache
+
+  def __call__(self, x, request_id: str):
+    if request_id not in self.caches:
+      self.init_cache(request_id)
+    else:
+      self.caches.move_to_end(request_id)
```
Comment (AlexCheema) [inline exo/inference/mlx/sharded_model.py:37]: nice so it's an LRU cache

File: exo/inference/tinygrad/inference.py [hunk 3]
```diff
@@ -58,44 +58,34 @@ def build_transformer(model_path: Path, shard: Shard, model_size="8B", device=No
...
-    await self.ensure_shard(shard)
-    start_pos = json.loads(inference_state or "{}").get("start_pos", 0)
-    n_captured_toks = json.loads(inference_state or "{}").get("n_captured_toks", 0)
-
-    toks = await asyncio.get_event_loop().run_in_executor(self.executor, self.tokenizer.encode, prompt)
-    h = await asyncio.get_event_loop().run_in_executor(self.executor, lambda: self.model(Tensor([toks]), start_pos, TEMPERATURE).realize())
+  async def sample(self, x: np.ndarray) -> np.ndarray:
+    logits = x[:, -1, :]
+    def sample_wrapper():
+      return sample_logits(Tensor(logits).flatten(), TEMPERATURE, 0, 0.8, 0.0, 0.0).realize()
```
Comment (AlexCheema) [inline exo/inference/tinygrad/inference.py:70]: not necessary in this PR but I think now that we have sampling happening at this level, we could make TEMPERATURE, TOP_K and whatever other parameters for sampling something that's configurable. This is actually part of the ChatGPT-API-Spec: you can pass temperature, but right now it doesn't do anything lol

File: exo/inference/tinygrad/models/llama.py [hunk 4]
```diff
@@ -201,32 +206,46 @@ def __init__(
     self.freqs_cis = precompute_freqs_cis(dim // n_heads, self.max_context*2, rope_theta, rope_scaling=rope_scaling).contiguous()
     self.forward_jit = TinyJit(self.forward) if jit else None
     self.shard = shard
+    self.caches = OrderedDict()
```
Comment (AlexCheema) [inline exo/inference/tinygrad/models/llama.py:209]: Is this a hack? I didn't think the caches would be in the model implementation itself?

File: exo/orchestration/standard_node.py [hunk 5]
```diff
@@ -258,49 +277,48 @@ async def forward_to_next_shard(
...
+          raise ValueError(f"Peer for {next_partition} not found")
+        
+        if is_tensor:
+          if DEBUG >= 1: print(f"Sending tensor to {target_peer.id()}: {tensor_or_prompt}")
+          await target_peer.send_tensor(next_shard, tensor_or_prompt, request_id=request_id, inference_state=inference_state)
+        else:
+          await target_peer.send_prompt(next_shard, tensor_or_prompt, request_id=request_id, inference_state=inference_state)
 
-  def get_current_shard(self, base_shard: Shard) -> Shard:
+  def get_partition_index(self, offset: int = 0):
```
Comment (AlexCheema) [inline exo/orchestration/standard_node.py:309]: offset is a future thing you're thinking about?

**Review by AlexCheema — CHANGES_REQUESTED** [review]
This looks great. Just take a Quick Look through my comments - there's 1 or 2 actionable things.

**Comment by AlexCheema** [comment]
beast mode.

**Pull request merged**